File: backend-node/src/routes/meal.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import aiService from '../services/aiService.js';
import { UserSession, Message, NutritionInfo, FoodItem } from '../database/models/index.js';

const router = express.Router();

// Analyze meal endpoint
router.post('/analyze-meal', async (req, res) => {
  try {
    const { message, language = 'auto', session_id } = req.body;
    
    if (!message) {
      return res.status(400).json({
        error: 'Message is required'
      });
    }
    
    // Get or create session
    let sessionId = session_id || uuidv4();
    let session = await UserSession.findByPk(sessionId);
    
    if (!session) {
      session = await UserSession.create({
        id: sessionId,
        userId: null,
        metadata: {}
      });
    }
    
    // Update last activity
    session.lastActivity = new Date();
    await session.save();
    
    // Store user message
    const userMessage = await Message.create({
      sessionId: session.id,
      role: 'user',
      content: message,
      metadata: {}
    });
    
    // Analyze meal using AI
    const aiResult = await aiService.analyzeMeal(message, language, session.id);
    
    // Create nutrition info if present
    let nutritionInfo = null;
    if (aiResult.nutrition && aiResult.nutrition.total_calories > 0) {
      nutritionInfo = await NutritionInfo.create({
        totalCalories: aiResult.nutrition.total_calories,
        totalProtein: aiResult.nutrition.total_protein,
        totalCarbs: aiResult.nutrition.total_carbs,
        totalFat: aiResult.nutrition.total_fat,
        totalFiber: aiResult.nutrition.total_fiber,
        totalSugar: aiResult.nutrition.total_sugar,
        totalSodium: aiResult.nutrition.total_sodium,
        analysisNotes: aiResult.analysis_notes
      });
      
      // Create food items
      if (aiResult.nutrition.food_items && aiResult.nutrition.food_items.length > 0) {
        for (const item of aiResult.nutrition.food_items) {
          await FoodItem.create({
            nutritionInfoId: nutritionInfo.id,
            name: item.name,
            nameCn: item.name_cn,
            amount: item.amount,
            unit: item.unit,
            calories: item.calories,
            protein: item.protein,
            carbs: item.carbs,
            fat: item.fat,
            fiber: item.fiber,
            sugar: item.sugar,
            sodium: item.sodium
          });
        }
      }
    }
    
    // Store assistant message
    const assistantMessage = await Message.create({
      sessionId: session.id,
      role: 'assistant',
      content: aiResult.ai_response,
      nutritionDataId: nutritionInfo ? nutritionInfo.id : null,
      metadata: {}
    });
    
    // Prepare response
    res.json({
      message_id: assistantMessage.id,
      nutrition: aiResult.nutrition,
      ai_response: aiResult.ai_response,
      session_id: session.id,
      timestamp: new Date().toISOString()
    });
    
  } catch (error) {
    console.error('Error analyzing meal:', error);
    res.status(500).json({
      error: 'Failed to analyze meal',
      message: error.message
    });
  }
});

export default router;

File: backend/app/schemas/meal.py
"""Meal analysis related schemas."""

from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime


class MealAnalysisRequest(BaseModel):
    """Request for meal analysis."""
    message: str = Field(..., min_length=1, max_length=5000, description="Meal description")
    session_id: Optional[str] = Field(None, description="Session ID for tracking")
    language: Optional[str] = Field("auto", description="Language preference (auto, en, zh)")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "message": "I had 2 scrambled eggs, 2 slices of whole wheat toast with butter, and a glass of orange juice",
            "session_id": "optional-session-id",
            "language": "en"
        }
    })


class FoodItemSchema(BaseModel):
    """Individual food item with nutrition data."""
    name: str = Field(..., description="Food item name")
    name_cn: Optional[str] = Field(None, description="Chinese name if applicable")
    amount: str = Field(..., description="Quantity/portion size")
    unit: Optional[str] = Field(None, description="Measurement unit")
    calories: float = Field(..., ge=0, description="Calories")
    protein: float = Field(..., ge=0, description="Protein in grams")
    carbs: float = Field(..., ge=0, description="Carbohydrates in grams")
    fat: float = Field(..., ge=0, description="Fat in grams")
    fiber: Optional[float] = Field(None, ge=0, description="Fiber in grams")
    sugar: Optional[float] = Field(None, ge=0, description="Sugar in grams")
    sodium: Optional[float] = Field(None, ge=0, description="Sodium in milligrams")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "name": "Scrambled Eggs",
            "amount": "2",
            "unit": "large eggs",
            "calories": 140,
            "protein": 12,
            "carbs": 1,
            "fat": 10
        }
    })


class NutritionInfoSchema(BaseModel):
    """Aggregated nutrition information."""
    total_calories: float = Field(..., ge=0, description="Total calories")
    total_protein: float = Field(..., ge=0, description="Total protein in grams")
    total_carbs: float = Field(..., ge=0, description="Total carbohydrates in grams")
    total_fat: float = Field(..., ge=0, description="Total fat in grams")
    total_fiber: Optional[float] = Field(None, ge=0, description="Total fiber in grams")
    total_sugar: Optional[float] = Field(None, ge=0, description="Total sugar in grams")
    total_sodium: Optional[float] = Field(None, ge=0, description="Total sodium in milligrams")
    food_items: List[FoodItemSchema] = Field(..., description="Individual food items")
    analysis_notes: Optional[str] = Field(None, description="Nutritional analysis and advice")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "total_calories": 450,
            "total_protein": 15.5,
            "total_carbs": 45.2,
            "total_fat": 18.3,
            "food_items": [],
            "analysis_notes": "This meal provides a good balance of protein and carbohydrates. Consider adding more vegetables for additional fiber and nutrients."
        }
    })


class MealAnalysisResponse(BaseModel):
    """Response from meal analysis."""
    message_id: str = Field(..., description="Unique message ID")
    nutrition: NutritionInfoSchema = Field(..., description="Nutrition breakdown")
    ai_response: str = Field(..., description="AI assistant's response")
    session_id: str = Field(..., description="Session ID")
    timestamp: datetime = Field(default_factory=datetime.utcnow, description="Analysis timestamp")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "message_id": "uuid-here",
            "nutrition": {},
            "ai_response": "I've analyzed your meal. You consumed approximately 450 calories with a good balance of macronutrients.",
            "session_id": "session-uuid",
            "timestamp": "2024-01-01T12:00:00Z"
        }
    })